Flakiness in `test_refund_transfer`

We had a [build]( https://circleci.com/gh/raiden-network/raiden/31024#artifacts/containers/0) failing with an assertion error in `test_refund_transfer`, here's the [log](https://31024-50646767-gh.circle-artifacts.com/0/test-reports/test-integration-matrix-3.7/results.xml) from the run : 

```
    def test_refund_transfer(
            raiden_chain,
            number_of_nodes,
            token_addresses,
            deposit,
            network_wait,
            retry_timeout,
            # UDP does not seem to retry messages until processed
            # https://github.com/raiden-network/raiden/issues/3185
            skip_if_not_matrix,
    ):
        """A failed transfer must send a refund back.
    
        TODO:
            - Unlock the token on refund #1091
            - Clear the merkletree and update the locked amount #193
            - Remove the refund message type #490"""
        # Topology:
        #
        #  0 -> 1 -> 2
        #
        app0, app1, app2 = raiden_chain
        token_address = token_addresses[0]
        payment_network_identifier = app0.raiden.default_registry.address
        token_network_identifier = views.get_token_network_identifier_by_token_address(
            views.state_from_app(app0),
            payment_network_identifier,
            token_address,
        )
    
        # make a transfer to test the path app0 -> app1 -> app2
        identifier_path = 1
        amount_path = 1
        mediated_transfer(
            app0,
            app2,
            token_network_identifier,
            amount_path,
            identifier_path,
            timeout=network_wait * number_of_nodes,
        )
    
        # drain the channel app1 -> app2
        identifier_drain = 2
        amount_drain = deposit * 8 // 10
        mediated_transfer(
            initiator_app=app1,
            target_app=app2,
            token_network_identifier=token_network_identifier,
            amount=amount_drain,
            identifier=identifier_drain,
            timeout=network_wait,
        )
    
        # wait for the nodes to sync
        gevent.sleep(0.2)
    
        assert_synced_channel_state(
            token_network_identifier,
            app0, deposit - amount_path, [],
            app1, deposit + amount_path, [],
        )
        assert_synced_channel_state(
            token_network_identifier,
            app1, deposit - amount_path - amount_drain, [],
>           app2, deposit + amount_path + amount_drain, [],
        )

raiden/tests/integration/transfer/test_refundtransfer.py:165: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
raiden/tests/utils/transfer.py:124: in assert_synced_channel_state
    assert_balance(channel1, balance1, locked_amount1)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

from_channel = <NettingChannelState id:1 opened:<TransactionExecutionStatus started:None finished:28 result:success> closed:None settled:None updated:None>
balance = 361, locked = 0

    def assert_balance(from_channel, balance, locked):
        """ Assert the from_channel overall token values. """
        assert balance >= 0
        assert locked >= 0
    
        distributable = balance - locked
        channel_distributable = channel.get_distributable(
            from_channel.our_state,
            from_channel.partner_state,
        )
    
>       assert channel.get_balance(from_channel.our_state, from_channel.partner_state) == balance
E       AssertionError

raiden/tests/utils/transfer.py:200: AssertionError
```

--- a/raiden/tests/integration/long_running/test_integration_events.py
+++ b/raiden/tests/integration/long_running/test_integration_events.py
@@ -20,7 +20,7 @@ from raiden.settings import DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS
 from raiden.tests.utils.events import must_have_event, search_for_item, wait_for_state_change
 from raiden.tests.utils.network import CHAIN
 from raiden.tests.utils.protocol import HoldOffChainSecretRequest
-from raiden.tests.utils.transfer import assert_synced_channel_state, get_channelstate
+from raiden.tests.utils.transfer import assert_synced_channel_state, get_channelstate, wait_assert
 from raiden.transfer import views
 from raiden.transfer.events import ContractSendChannelClose
 from raiden.transfer.mediated_transfer.events import SendLockedTransfer
@@ -241,12 +241,13 @@ def test_channel_deposit(raiden_chain, deposit, retry_timeout, token_addresses):
     )
 
     wait_both_channel_open(app0, app1, registry_address, token_address, retry_timeout)
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, 0, [],
-        app1, 0, [],
-    )
+    with gevent.Timeout(retry_timeout):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, 0, [],
+            app1, 0, [],
+        )
 
     RaidenAPI(app0.raiden).set_total_channel_deposit(
         registry_address,
@@ -263,12 +264,13 @@ def test_channel_deposit(raiden_chain, deposit, retry_timeout, token_addresses):
         deposit,
         retry_timeout,
     )
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, 0, [],
-    )
+    with gevent.Timeout(retry_timeout):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, 0, [],
+        )
 
     RaidenAPI(app1.raiden).set_total_channel_deposit(
         registry_address,
@@ -285,12 +287,13 @@ def test_channel_deposit(raiden_chain, deposit, retry_timeout, token_addresses):
         deposit,
         retry_timeout,
     )
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, deposit, [],
-    )
+    with gevent.Timeout(retry_timeout):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, deposit, [],
+        )
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])
@@ -402,11 +405,13 @@ def test_query_events(
     )
     assert channelcount0 + 1 == channelcount1
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, 0, [],
-        app1, 0, [],
-    )
+    with gevent.Timeout(retry_timeout):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, 0, [],
+            app1, 0, [],
+        )
 
     RaidenAPI(app0.raiden).set_total_channel_deposit(
         registry_address,
@@ -575,11 +580,13 @@ def test_secret_revealed_on_chain(
 
     # The channel app0-app1 should continue with the protocol off-chain, once
     # the secret is released on-chain by app2
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount, [],
-        app1, deposit + amount, [],
-    )
+    with gevent.Timeout(10):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount, [],
+            app1, deposit + amount, [],
+        )
 
     with gevent.Timeout(10):
         wait_for_state_change(

--- a/raiden/tests/integration/long_running/test_settlement.py
+++ b/raiden/tests/integration/long_running/test_settlement.py
@@ -17,6 +17,7 @@ from raiden.tests.utils.transfer import (
     assert_synced_channel_state,
     get_channelstate,
     mediated_transfer,
+    wait_assert,
 )
 from raiden.transfer import channel, views
 from raiden.transfer.state import UnlockProofState
@@ -128,7 +129,7 @@ def test_settle_is_automatically_called(raiden_network, token_addresses):
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])
-def test_lock_expiry(raiden_network, token_addresses, deposit):
+def test_lock_expiry(raiden_network, token_addresses, deposit, network_wait):
     """Test lock expiry and removal."""
     alice_app, bob_app = raiden_network
     token_address = token_addresses[0]
@@ -194,11 +195,13 @@ def test_lock_expiry(raiden_network, token_addresses, deposit):
     #    A -> B LockedTransfer
     #    B -> A SecretRequest
     #    - protocol didn't continue
-    assert_synced_channel_state(
-        token_network_identifier,
-        alice_app, deposit, [lock],
-        bob_app, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            alice_app, deposit, [lock],
+            bob_app, deposit, [],
+        )
 
     # Verify lock is registered in both channel states
     alice_channel_state = get_channelstate(alice_app, bob_app, token_network_identifier)
@@ -252,6 +255,7 @@ def test_batch_unlock(
         secret_registry_address,
         deposit,
         blockchain_type,
+        network_wait,
 ):
     """Batch unlock can be called after the channel is settled."""
     alice_app, bob_app = raiden_network
@@ -313,11 +317,13 @@ def test_batch_unlock(
     #    A -> B LockedTransfer
     #    B -> A SecretRequest
     #    - protocol didn't continue
-    assert_synced_channel_state(
-        token_network_identifier,
-        alice_app, deposit, [lock],
-        bob_app, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            alice_app, deposit, [lock],
+            bob_app, deposit, [],
+        )
 
     # Take a snapshot early on
     alice_app.raiden.wal.snapshot()

--- a/raiden/tests/integration/test_pythonapi.py
+++ b/raiden/tests/integration/test_pythonapi.py
@@ -28,6 +28,7 @@ from raiden.tests.utils.transfer import (
     assert_synced_channel_state,
     get_channelstate,
     mediated_transfer,
+    wait_assert,
 )
 from raiden.transfer import views
 from raiden.transfer.events import EventPaymentReceivedSuccess, EventPaymentSentSuccess
@@ -265,7 +266,7 @@ def test_transfer_to_unknownchannel(raiden_network, token_addresses):
 @pytest.mark.parametrize('number_of_nodes', [2])
 @pytest.mark.parametrize('number_of_tokens', [2])
 @pytest.mark.parametrize('settle_timeout', [TEST_TOKEN_SWAP_SETTLE_TIMEOUT])
-def test_token_swap(raiden_network, deposit, token_addresses):
+def test_token_swap(raiden_network, deposit, token_addresses, network_wait):
     app0, app1 = raiden_network
 
     maker_address = app0.raiden.address
@@ -299,19 +300,21 @@ def test_token_swap(raiden_network, deposit, token_addresses):
     assert async_result.wait()
 
     # wait for the taker to receive and process the messages
-    gevent.sleep(0.5)
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            maker_token,
+            app0, deposit - maker_amount, [],
+            app1, deposit + maker_amount, [],
+        )
 
-    assert_synced_channel_state(
-        maker_token,
-        app0, deposit - maker_amount, [],
-        app1, deposit + maker_amount, [],
-    )
-
-    assert_synced_channel_state(
-        taker_token,
-        app0, deposit + taker_amount, [],
-        app1, deposit - taker_amount, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            taker_token,
+            app0, deposit + taker_amount, [],
+            app1, deposit - taker_amount, [],
+        )
 
 
 @pytest.mark.parametrize('channels_per_node', [1])

--- a/raiden/tests/integration/test_recovery.py
+++ b/raiden/tests/integration/test_recovery.py
@@ -1,3 +1,5 @@
+import gevent
+
 import pytest
 from gevent import server
 
@@ -11,7 +13,7 @@ from raiden.raiden_event_handler import RaidenEventHandler
 from raiden.tests.utils.events import search_for_item
 from raiden.tests.utils.network import CHAIN
 from raiden.tests.utils.protocol import WaitForMessage
-from raiden.tests.utils.transfer import assert_synced_channel_state, mediated_transfer
+from raiden.tests.utils.transfer import assert_synced_channel_state, mediated_transfer, wait_assert
 from raiden.transfer import views
 from raiden.transfer.state_change import (
     ContractReceiveChannelClosed,
@@ -91,17 +93,20 @@ def test_recovery_happy_case(
     del app0  # from here on the app0_restart should be used
 
     app0_restart.start()
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0_restart, deposit - spent_amount, [],
-        app1, deposit + spent_amount, [],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - spent_amount, [],
-        app2, deposit + spent_amount, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0_restart, deposit - spent_amount, [],
+            app1, deposit + spent_amount, [],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - spent_amount, [],
+            app2, deposit + spent_amount, [],
+        )
 
     # wait for the nodes' healthcheck to update the network statuses
     waiting.wait_for_healthy(
@@ -134,18 +139,22 @@ def test_recovery_happy_case(
         timeout=network_wait * number_of_nodes * 2,
     )
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0_restart, deposit - spent_amount, [],
-        app1, deposit + spent_amount, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0_restart, deposit - spent_amount, [],
+            app1, deposit + spent_amount, [],
+        )
 
     wait_for_payment.wait()
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - spent_amount, [],
-        app2, deposit + spent_amount, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - spent_amount, [],
+            app2, deposit + spent_amount, [],
+        )
 
 
 @pytest.mark.parametrize('deposit', [10])

--- a/raiden/tests/integration/test_send_queued_messages.py
+++ b/raiden/tests/integration/test_send_queued_messages.py
@@ -9,7 +9,7 @@ from raiden.raiden_event_handler import RaidenEventHandler
 from raiden.tests.utils.events import raiden_events_search_for_item
 from raiden.tests.utils.network import CHAIN
 from raiden.tests.utils.protocol import HoldRaidenEvent, dont_handle_node_change_network_state
-from raiden.tests.utils.transfer import assert_synced_channel_state
+from raiden.tests.utils.transfer import assert_synced_channel_state, wait_assert
 from raiden.transfer import views
 from raiden.transfer.events import EventPaymentSentSuccess
 from raiden.transfer.mediated_transfer.events import SendSecretReveal
@@ -124,11 +124,13 @@ def test_send_queued_messages(
         retry_timeout=network_wait,
     )
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0_restart, deposit - spent_amount, [],
-        app1, deposit + spent_amount, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0_restart, deposit - spent_amount, [],
+            app1, deposit + spent_amount, [],
+        )
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])

--- a/raiden/tests/integration/transfer/test_mediatedtransfer_invalid.py
+++ b/raiden/tests/integration/transfer/test_mediatedtransfer_invalid.py
@@ -30,6 +30,7 @@ def test_failsfast_lockedtransfer_exceeding_distributable(
         raiden_network,
         token_addresses,
         deposit,
+        network_wait,
 ):
 
     app0, app1 = raiden_network
@@ -51,11 +52,13 @@ def test_failsfast_lockedtransfer_exceeding_distributable(
     assert payment_status.payment_done.get(timeout=5) is False
     assert payment_status.payment_done.successful()
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, deposit, [],
+        )
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])
@@ -154,6 +157,7 @@ def test_receive_lockedtransfer_invalidsender(
         token_addresses,
         deposit,
         reveal_timeout,
+        network_wait,
 ):
 
     app0, app1 = raiden_network
@@ -191,12 +195,13 @@ def test_receive_lockedtransfer_invalidsender(
         LocalSigner(other_key),
         app0,
     )
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, deposit, [],
+        )
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])
@@ -206,6 +211,7 @@ def test_receive_lockedtransfer_invalidrecipient(
         token_addresses,
         reveal_timeout,
         deposit,
+        network_wait,
 ):
 
     app0, app1 = raiden_network
@@ -245,11 +251,13 @@ def test_receive_lockedtransfer_invalidrecipient(
         app1,
     )
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, deposit, [],
+        )
 
 
 @pytest.mark.parametrize('number_of_nodes', [2])
@@ -260,6 +268,7 @@ def test_received_lockedtransfer_closedchannel(
         reveal_timeout,
         token_addresses,
         deposit,
+        network_wait,
 ):
 
     app0, app1 = raiden_network
@@ -311,8 +320,10 @@ def test_received_lockedtransfer_closedchannel(
     )
 
     # The local state must not change since the channel is already closed
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [],
-        app1, deposit, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [],
+            app1, deposit, [],
+        )

--- a/raiden/tests/integration/transfer/test_refundtransfer.py
+++ b/raiden/tests/integration/transfer/test_refundtransfer.py
@@ -17,6 +17,7 @@ from raiden.tests.utils.transfer import (
     assert_synced_channel_state,
     get_channelstate,
     mediated_transfer,
+    wait_assert,
 )
 from raiden.transfer import channel, views
 from raiden.transfer.mediated_transfer.events import (
@@ -33,7 +34,7 @@ from raiden.waiting import wait_for_block, wait_for_settle
 @pytest.mark.parametrize('channels_per_node', [CHAIN])
 @pytest.mark.parametrize('number_of_nodes', [3])
 @pytest.mark.parametrize('settle_timeout', [50])
-def test_refund_messages(raiden_chain, token_addresses, deposit):
+def test_refund_messages(raiden_chain, token_addresses, deposit, network_wait):
     # The network has the following topology:
     #
     #   App0 <---> App1 <---> App2
@@ -80,18 +81,22 @@ def test_refund_messages(raiden_chain, token_addresses, deposit):
     send_refundtransfer = raiden_events_search_for_item(app1.raiden, SendRefundTransfer, {})
     assert send_refundtransfer
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit, [send_lockedtransfer.transfer.lock],
-        app1, deposit, [send_refundtransfer.transfer.lock],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit, [send_lockedtransfer.transfer.lock],
+            app1, deposit, [send_refundtransfer.transfer.lock],
+        )
 
     # This channel was exhausted to force the refund transfer
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, 0, [],
-        app2, deposit * 2, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, 0, [],
+            app2, deposit * 2, [],
+        )
 
 
 @pytest.mark.parametrize('privatekey_seed', ['test_refund_transfer:{}'])
@@ -152,18 +157,22 @@ def test_refund_transfer(
     )
 
     # wait for the nodes to sync
-    gevent.sleep(0.2)
+    gevent.sleep(1)
 
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [],
-        app1, deposit + amount_path, [],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path - amount_drain, [],
-        app2, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [],
+            app1, deposit + amount_path, [],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path - amount_drain, [],
+            app2, deposit + amount_path + amount_drain, [],
+        )
 
     # app0 -> app1 -> app2 is the only available path, but the channel app1 ->
     # app2 doesn't have capacity, so a refund will be sent on app1 -> app0
@@ -201,16 +210,20 @@ def test_refund_transfer(
     assert lock.secrethash == refund_lock.secrethash
 
     # Both channels have the amount locked because of the refund message
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [lock],
-        app1, deposit + amount_path, [refund_lock],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path - amount_drain, [],
-        app2, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [lock],
+            app1, deposit + amount_path, [refund_lock],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path - amount_drain, [],
+            app2, deposit + amount_path + amount_drain, [],
+        )
 
     # Additional checks for LockExpired causing nonce mismatch after refund transfer:
     # https://github.com/raiden-network/raiden/issues/3146#issuecomment-447378046
@@ -340,19 +353,20 @@ def test_different_view_of_last_bp_during_unlock(
         timeout=network_wait,
     )
 
-    # wait for the nodes to sync
-    gevent.sleep(0.2)
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [],
-        app1, deposit + amount_path, [],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path - amount_drain, [],
-        app2, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [],
+            app1, deposit + amount_path, [],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path - amount_drain, [],
+            app2, deposit + amount_path + amount_drain, [],
+        )
 
     # app0 -> app1 -> app2 is the only available path, but the channel app1 ->
     # app2 doesn't have capacity, so a refund will be sent on app1 -> app0
@@ -390,16 +404,20 @@ def test_different_view_of_last_bp_during_unlock(
     assert lock.secrethash == refund_lock.secrethash
 
     # Both channels have the amount locked because of the refund message
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [lock],
-        app1, deposit + amount_path, [refund_lock],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path - amount_drain, [],
-        app2, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [lock],
+            app1, deposit + amount_path, [refund_lock],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path - amount_drain, [],
+            app2, deposit + amount_path + amount_drain, [],
+        )
 
     # Additional checks for LockExpired causing nonce mismatch after refund transfer:
     # https://github.com/raiden-network/raiden/issues/3146#issuecomment-447378046
@@ -531,24 +549,27 @@ def test_refund_transfer_after_2nd_hop(
         timeout=network_wait,
     )
 
-    # wait for the nodes to sync
-    gevent.sleep(0.2)
-
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [],
-        app1, deposit + amount_path, [],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path, [],
-        app2, deposit + amount_path, [],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app2, deposit - amount_path - amount_drain, [],
-        app3, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [],
+            app1, deposit + amount_path, [],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path, [],
+            app2, deposit + amount_path, [],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app2, deposit - amount_path - amount_drain, [],
+            app3, deposit + amount_path + amount_drain, [],
+        )
 
     # app0 -> app1 -> app2 > app3 is the only available path, but the channel
     # app2 -> app3 doesn't have capacity, so a refund will be sent on
@@ -600,18 +621,24 @@ def test_refund_transfer_after_2nd_hop(
     assert lock2.expiration
 
     # channels have the amount locked because of the refund message
-    assert_synced_channel_state(
-        token_network_identifier,
-        app0, deposit - amount_path, [lock1],
-        app1, deposit + amount_path, [refund_lock1],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app1, deposit - amount_path, [lock2],
-        app2, deposit + amount_path, [refund_lock2],
-    )
-    assert_synced_channel_state(
-        token_network_identifier,
-        app2, deposit - amount_path - amount_drain, [],
-        app3, deposit + amount_path + amount_drain, [],
-    )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app0, deposit - amount_path, [lock1],
+            app1, deposit + amount_path, [refund_lock1],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app1, deposit - amount_path, [lock2],
+            app2, deposit + amount_path, [refund_lock2],
+        )
+    with gevent.Timeout(network_wait):
+        wait_assert(
+            assert_synced_channel_state,
+            token_network_identifier,
+            app2, deposit - amount_path - amount_drain, [],
+            app3, deposit + amount_path + amount_drain, [],
+        )

